rust-app/zephyr/src/uart.rs

```diff
@@ -46,10 +52,95 @@ macro_rules! trait_impl {
                     zephyr_sys::syscalls::$context::uart_err_check(core::mem::transmute(device))
                 })
             }
+
+            #[cfg(feature = "have_std")]
+            #[inline(always)]
+            fn uart_config_get(device: &Device) -> Result<UartConfig, i32> {
+                let mut config = UartConfig::default();
+                let rc = unsafe {
+                    zephyr_sys::syscalls::user::uart_config_get(
+                        core::mem::transmute(device),
+                        &mut config.0,
+                    )
+                };
+                if rc == 0 {
+                    Ok(config)
+                } else {
+                    Err(rc)
+                }
+            }
+
+            #[cfg(feature = "have_std")]
+            #[inline(always)]
+            fn uart_configure(device: &Device, config: &UartConfig) -> i32 {
+                (unsafe {
+                    zephyr_sys::syscalls::user::uart_configure(
+                        core::mem::transmute(device),
+                        &config.0,
+                    )
+                })
+            }
         }
     };
 }
 
+#[cfg(feature = "have_std")]
+pub struct UartConfig(zephyr_sys::raw::uart_config);
+
+#[cfg(feature = "have_std")]
+impl UartConfig {
+    pub fn set_flow_control_rts_cts(&mut self) {
+        self.0.flow_ctrl =
+            zephyr_sys::raw::uart_config_flow_control_UART_CFG_FLOW_CTRL_RTS_CTS as u8;
+    }
+
+    pub fn set_flow_control_dtr_dsr(&mut self) {
+        self.0.flow_ctrl =
+            zephyr_sys::raw::uart_config_flow_control_UART_CFG_FLOW_CTRL_DTR_DSR as u8;
+    }
+
+    pub fn disable_flow_control(&mut self) {
+        self.0.flow_ctrl = zephyr_sys::raw::uart_config_flow_control_UART_CFG_FLOW_CTRL_NONE as u8;
+    }
+
+    pub fn get_baud_rate(&self) -> u32 {
+        self.0.baudrate
+    }
+
+    pub fn set_baud_rate(&mut self, baud_rate: u32) {
+        self.0.baudrate = baud_rate;
+    }
+
+    pub fn get_stop_bits(&self) -> u8 {
+        self.0.stop_bits
+    }
+
+    pub fn set_stop_bits(&mut self, stop_bits: u8) {
+        self.0.stop_bits = stop_bits
+    }
+
+    pub fn get_data_bits(&self) -> u8 {
+        self.0.data_bits
+    }
+
+    pub fn set_data_bits(&mut self, data_bits: u8) {
+        self.0.data_bits = data_bits
+    }
+}
+
+#[cfg(feature = "have_std")]
```

Comment:
I don't think we need the have_std guard for the uart API. Yeah, you can't use device_get_binding, but if you had a device reference any other way the uart API doesn't need cstr.